File: C11/백준/Silver/30804. 과일 탕후루/과일 탕후루.c
#include <stdio.h>

int max(int a, int b){
	return a > b ? a : b;
}

int main(void){
	
	int N;
	scanf("%d", &N);
	
	int fruits[N];
	for(int i = 0; i < N; i++){
		scanf("%d", &fruits[i]);
	}
	
	int count[10] = {0};
	int left = 0, right = 0;
	int kind = 0;
	int maxLength = 0;
	
	while(right < N){
		if(count[fruits[right]] == 0) kind++;
		count[fruits[right]]++;
		right++;
		
		while(kind > 2){
			count[fruits[left]]--;
            if (count[fruits[left]] == 0) kind--;
            left++;
		}
		
		maxLength = max(maxLength, right - left);
	}
	
	printf("%d", maxLength);
	
	return 0;
}

File: Java/백준/Silver/30804. 과일 탕후루/과일 탕후루.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        Integer[] fruits = new Integer[N];
        String[] line = br.readLine().split(" ");

        for (int i = 0; i < N; i++) {
            fruits[i] = Integer.parseInt(line[i]);
        }

        int[] count = new int[10];
        int left = 0, right = 0;
        int kind = 0;
        int maxLength = 0;

        while (right < N) {
            if (count[fruits[right]] == 0) kind++;
            count[fruits[right]]++;
            right++;

            while (kind > 2) {
                count[fruits[left]]--;
                if (count[fruits[left]] == 0) kind--;
                left++;
            }

            maxLength = Math.max(maxLength, right - left);
        }

        System.out.print(maxLength);
    }
}